src/audio/module_adapter/module/volume/volume_ipc4.c

```diff
@@ -0,0 +1,419 @@
+
+	ret = module_set_configuration(mod, config_id, pos, data_offset_size, fragment,
+				       fragment_size, response, response_size);
+	if (ret < 0)
+		return ret;
+
+	/* return if more fragments are expected or if the module is not prepared */
+	if (pos != MODULE_CFG_FRAGMENT_LAST && pos != MODULE_CFG_FRAGMENT_SINGLE)
+		return 0;
+
+	switch (config_id) {
+	case IPC4_VOLUME:
+		return volume_set_volume(mod, fragment, fragment_size);
+	case IPC4_SET_ATTENUATION:
+		return volume_set_attenuation(mod, fragment, fragment_size);
+	default:
+		comp_err(dev, "unsupported param %d", config_id);
+		return -EINVAL;
+	}
+
+	return 0;
+}
+
+int volume_get_config(struct processing_module *mod,
+		      uint32_t config_id, uint32_t *data_offset_size,
+		      uint8_t *fragment, size_t fragment_size)
+{
+	struct vol_data *cd = module_get_private_data(mod);
+	struct ipc4_peak_volume_config *cdata;
+	int i;
+
+	comp_dbg(mod->dev, "volume_get_large_config()");
+
+	cdata = (struct ipc4_peak_volume_config *)ASSUME_ALIGNED(fragment, 8);
+
+	switch (config_id) {
+	case IPC4_VOLUME:
+		for (i = 0; i < cd->channels; i++) {
+			uint32_t volume = cd->peak_regs.target_volume[i];
+
+			cdata[i].channel_id = i;
+			cdata[i].target_volume = convert_volume_ipc3_to_ipc4(volume);
+		}
+		*data_offset_size = sizeof(*cdata) * cd->channels;
+		break;
+	default:
+		comp_err(mod->dev, "unsupported param %d", config_id);
+		return -EINVAL;
+	}
+
+	return 0;
+}
+
+static int volume_params(struct processing_module *mod)
+{
+	struct sof_ipc_stream_params *params = mod->stream_params;
+	struct comp_buffer __sparse_cache *sink_c, *source_c;
+	struct comp_buffer *sinkb, *sourceb;
+	struct comp_dev *dev = mod->dev;
+
+	comp_dbg(dev, "volume_params()");
+
+	ipc4_base_module_cfg_to_stream_params(&mod->priv.cfg.base_cfg, params);
+
+	component_set_nearest_period_frames(dev, params->rate);
+
+	/* volume component will only ever have 1 sink buffer */
+	sinkb = list_first_item(&dev->bsink_list, struct comp_buffer, source_list);
+	sink_c = buffer_acquire(sinkb);
+	ipc4_update_buffer_format(sink_c, &mod->priv.cfg.base_cfg.audio_fmt);
+	buffer_release(sink_c);
+
+	sourceb = list_first_item(&dev->bsource_list, struct comp_buffer, sink_list);
+	source_c = buffer_acquire(sourceb);
+	ipc4_update_buffer_format(source_c, &mod->priv.cfg.base_cfg.audio_fmt);
+	buffer_release(source_c);
+
+	return 0;
+}
+
+void volume_update_current_vol_ipc4(struct vol_data *cd)
+{
+	int i;
+
+	assert(cd);
+
+	for (i = 0; i < cd->channels; i++)
+		cd->peak_regs.current_volume[i] = cd->volume[i];
+}
+
+int volume_peak_prepare(struct vol_data *cd, struct processing_module *mod)
+{
+	int ret;
+
+	cd->peak_cnt = 0;
+#if CONFIG_COMP_PEAK_VOL
+	cd->peak_report_cnt = CONFIG_PEAK_METER_UPDATE_PERIOD * 1000 / mod->dev->period;
+	if (cd->peak_report_cnt == 0)
+		cd->peak_report_cnt = 1;
+#endif
+	ret = volume_params(mod);
+	if (ret < 0)
+		return ret;
+
+	return 0;
+}
```

Comment:
@btian1 just over 400 lines - looks much better to me now! Thanks for doing the work!